apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/TypeTests.cs

```diff
@@ -0,0 +1,643 @@
+            {
+                complexStruct, s_complexStruct,
+                "{ \"StructField\": { \"StringField\": \"stringValue\", \"Int64Field\": \"2\", \"Float64Field\": \"NaN\", \"BoolField\": true, \"DateField\": \"2017-01-31\", \"TimestampField\": \"2017-01-31T03:15:30Z\" }, \"ArrayField\": [ \"4\", \"5\", \"6\" ] }"
+            };
+        }
+
+        public static IEnumerable<object[]> GetInvalidValueConversions()
+        {
+            //Spanner type = Float64 tests.
+            yield return new object[] {(char) 1, SpannerDbType.Float64, ""};
+            yield return new object[] {s_testDate, SpannerDbType.Float64, ""};
+            yield return new object[] {new ToStringClass("1.5"), SpannerDbType.Float64, ""};
+            yield return new object[] {"", SpannerDbType.Float64, ""};
+
+            //Spanner type = Int64 tests.
+            yield return new object[] {s_testDate, SpannerDbType.Int64, ""};
+            yield return new object[] {double.NegativeInfinity, SpannerDbType.Int64, ""};
+            yield return new object[] {double.PositiveInfinity, SpannerDbType.Int64, ""};
+            yield return new object[] {double.NaN, SpannerDbType.Int64, ""};
+            yield return new object[] {"1.5", SpannerDbType.Int64, Quote("2")};
+            yield return new object[] {new ToStringClass("1.5"), SpannerDbType.Int64, ""};
+
+            //Spanner type = Bool tests.
+            yield return new object[] {(char) 1, SpannerDbType.Bool, ""};
+            yield return new object[] {"1", SpannerDbType.Bool, ""};
+            yield return new object[] {new ToStringClass("true"), SpannerDbType.Bool, ""};
+
+            //Spanner type = String tests.
+            //(all work)
+
+            //Spanner type = Date tests.
+            yield return new object[] {new ToStringClass("hello"), SpannerDbType.Date, ""};
+            yield return new object[] {"badjuju", SpannerDbType.Date, ""};
+        }
+
+        private string StrippedString(string s) => new StringBuilder(s)
+            .Replace("{", "")
+            .Replace("}", "")
+            .Replace("[", "")
+            .Replace("]", "")
+            .Replace(" ", "")
+            .Replace("\"StructField\":", "")
+            .ToString();
+
+        private bool IsEqualWhenDecomposed(string expected, string actual)
+        {
+            //There are cases where the serialization (of structs) is nondeterministic
+            //This would cause the unit tests to fail randomly, so if we see a failure
+            //we decompose the expected and actual into their respective pieces and validate
+            //each piece.
+            //we try to decompose the actual and expected in case its an unordered dictionary.
+            expected = StrippedString(expected);
+            actual = StrippedString(actual);
+            var unorderedMatchExpected = new HashSet<string>(expected.Split(','));
+            var unorderedMatchActual = new HashSet<string>(actual.Split(','));
```

Comment:
Did parsing as a struct not work, or is that still TBD? (You could always add a TODO to do that later, to get this in sooner.)